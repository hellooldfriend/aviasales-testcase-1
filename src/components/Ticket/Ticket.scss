@import '../../variables.scss';

.ticket {
    display: inline-block;
    width: 100%;
    margin-bottom: 20px;
    background: $background_secondary;
    box-sizing: border-box;
    padding: $p2;

    &-price {
        font-size: 24px;
        color: $color_secondary;
    }

    &-info {
        width: 100%;
        display: inline-flex;
        justify-content: space-between;
        margin-bottom: 20px;
    }

    &-segment {
        display: grid;
        grid-template-columns: repeat(3, 1fr);
        grid-column-gap: 30px;
        margin-bottom: 10px;

        &-title {
            color: $color_subtitle;
            text-transform: uppercase;
            line-height: 18px;
            letter-spacing: 0.5px;
        }

        &-body {
            font-size: 14px;
        }
    }

}