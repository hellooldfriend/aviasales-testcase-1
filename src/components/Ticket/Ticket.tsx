import React from 'react'
import './Ticket.scss'


import { ITicket, ITicketSegment } from '../../types'
import {
    formatDuration,
    getEndDatetime,
    formatTime,
    formatPrice,
    decline
} from '../../functions'


const Ticket = (ticket: ITicket) => {
    /*
    * Custom renders
    *
    *
    */
    function renderInfo() {
        return (
            <>
                <div className="ticket-price">
                    {formatPrice(ticket.price)}
                </div>

                <img
                    src={`http://pics.avs.io/99/36/${ticket.carrier}.png`}
                    alt={ticket.carrier}
                    style={{ width: '100px', height: '36px' }}
                />

            </>
        )
    }

    function renderSegment(segment: ITicketSegment) {
        return (
            <div
                key={`${segment.origin}_${segment.duration}_${segment.date}`}
                className="ticket-segment"
            >
                <div>
                    <div className="ticket-segment-title">
                        {segment.origin} – {segment.destination}
                    </div>
                    <div className="ticket-segment-body">
                        {formatTime(segment.date)} – {formatTime(getEndDatetime(segment.date, segment.duration))}
                    </div>
                </div>

                <div>
                    <div className="ticket-segment-title">
                        В пути
                    </div>
                    <div className="ticket-segment-body">
                        {formatDuration(segment.duration)}
                    </div>
                </div>

                {segment.stops.length > 0 &&
                    <div>
                        <div className="ticket-segment-title">
                            {segment.stops.length} {decline(segment.stops.length, ['пересадка', 'пересадки', 'пересадок'])}
                        </div>
                        <div className="ticket-segment-body">
                            {segment.stops.join(', ')}
                        </div>
                    </div>
                }

            </div>
        )
    }

    /*
   * Final render
   *
   *
   */
    return (
        <div className="ticket">
            <div className="ticket-info">
                {renderInfo()}
            </div>
            {ticket.segments.map(renderSegment)}
        </div>
    )
}


export default Ticket