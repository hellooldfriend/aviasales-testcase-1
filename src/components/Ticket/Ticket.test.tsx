import React from 'react'
import { render, screen, cleanup } from '@testing-library/react'
import renderer from 'react-test-renderer'

import Ticket from './Ticket'

afterEach(() => {
    cleanup()
})

const ticket = {
    price: '2 000',
    carrier: 'XXX',
    segments: [],
}

describe('Ticket', () => {
    test('Should render Ticket', () => {
        render(<Ticket {...ticket} />)
        const element = screen.getByTestId('ticket')
        expect(element).toBeInTheDocument()
        expect(element).toHaveTextContent(ticket.price)
    })

    test('Matches snapshot', () => {
        const element = renderer.create(
            <Ticket {...ticket} />
        ).toJSON()
        expect(element).toMatchSnapshot()
    })
})

