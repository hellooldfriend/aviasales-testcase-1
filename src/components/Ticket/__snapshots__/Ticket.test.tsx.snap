// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ticket Matches snapshot 1`] = `
<div
  className="ticket"
  data-testid="ticket"
>
  <div
    className="ticket-info"
  >
    <div
      className="ticket-price"
    >
      2 000
    </div>
    <img
      alt="XXX"
      src="http://pics.avs.io/99/36/XXX.png"
      style={
        Object {
          "height": "36px",
          "width": "100px",
        }
      }
    />
  </div>
  <div
    className="ticket-segment"
  >
    <div>
      <div
        className="ticket-segment-title"
      >
        YYY
         – 
        20
      </div>
      <div
        className="ticket-segment-body"
      >
        NaN:NaN
         – 
        NaN:NaN
      </div>
    </div>
    <div>
      <div
        className="ticket-segment-title"
      >
        В пути
      </div>
      <div
        className="ticket-segment-body"
      >
        1ч 40м
      </div>
    </div>
    <div>
      <div
        className="ticket-segment-title"
      >
        2
         
        пересадки
      </div>
      <div
        className="ticket-segment-body"
      >
        aaa, bbb
      </div>
    </div>
  </div>
</div>
`;
